salt/orch/update.sls

```diff
@@ -265,25 +286,46 @@ early-services-setup:
     - kwarg:
         destructive: True
     - require:
-      - {{ master_id }}-start-services
-
-{% endfor %}
+      - all-masters-post-start-services
+{%- endfor %}
 
-# Perform migrations after all masters have been updated
-all-masters-post-start-services:
+# NOTE: Remove me for 4.0
+#
+# On 2.0 -> 3.0 we are updating the way kubelets auth against the apiservers.
+# At this point in time all masters have been updated, and all workers are (or
+# will) be in `NotReady` state. This means that any operation that we perform
+# that go through the apiserver down to the kubelets won't work (e.g. draining
+# nodes).
+#
+# To fix this problem we'll apply the haproxy sls to all worker nodes, so they
+# can rejoin the cluster and we can operate on them normally.
+all-workers-2.0-pre-clean-shutdown:
   salt.state:
-    - tgt: '{{ is_master_tgt }}'
+    - tgt: '( {{ is_updateable_worker_tgt }} ) and G@osrelease:2.0'
```

Comment:
In case we have this grain `G@osrelease:2.0`, maybe we could protect all the blocks were we run `migration-2-3-*`...

Reply:
I only protected the one that would certainly not work, new haproxy on old system will break, as will do an old haproxy on a new system.

However, being on a 3.0 snapshot doesn't mean that we should not run a migration yet (e.g. if an update failed and we are retrying it, we still want to rename nodes, delete old node names, deploy cni...).

That is why I only explicitly set `G@osrelease:2.0` with the old `haproxy` and `G@osrelease:3.0` with the new haproxy.

Reply:
The thing is that we are introducing many `migrations` and I don't like them to be unconditionally executed: they are just run when the update orchestration is run. In some of them we are not effectively checking if the migration must be applied or not (I'm thinking for example on `etcd`). And I don't think those checks should be done individually in the `migration-*` files either.

So my personal take on this would be to find a general solution at the orchestration level that would run `migration-x-y`  _iff_ the node is at `x` and it is being updated to `y`.

Reply:
I think this logic is very probably out of the scope for this PR. I believe we should do something similar to what you suggest, but not right now IMHO (in this PR). What I did was to move all migration logic from code mixed with the normal states into a specific `migration` folder, nothing else.

In any case, I have been thinking a bit about this, so we have two kinds of migrations:

* Cluster-wide migrations
  * We cannot know if we ran a specific migration that affects the whole cluster (e.g. deploying PSP, or CNI new manifests).
    * We would need a mechanism where we could save this information (e.g. callbacks to a central piece like Velum that saves cluster-wide migrations), or
    * Always run the cluster-wide migrations; making them idempotent from the very beginning (as they are now).

* Node specific migrations
  * Migrations that only affect the current node.
  * Relying on `osrelease` is truly problematic and I'll detail why:
    * When `transactional-update` creates a new snapshot, it sets the default btrfs snapshot to the new one: https://github.com/openSUSE/transactional-update/blob/2b46aa83d6ec39e050a67876c5618ea11a4df27e/sbin/transactional-update.in#L956.
    * This means that an "innocent" reboot of a machine could upgrade it (loading a new snapshot) and we could skip migrations on that machine (because its `osrelease` is not what we expect to be, and we skip the migrations on that one).
  * Migrations could be identified with some string or hash and we could save the migrations that have been executed in the machine's grains as a list.
    * Ala-rails, we run the migrations that are not present in the machine list of migrations run. This is absolutely snapshot agnostic; either you have run a migration or not; if not: do it.
  * Aside from migrations, I think it would be good that we control when we `btrfs set-default` a new snapshot. If we could add a new flag to `transactional-update` that skips the call to `btrfs set-default` when it creates a new snapshot, then we could, when upgrading:
    * Get machines with `tx_update_reboot_needed` flag
    * For each machine, we run `btrfs subvolume set-default $BTRFS_ID ${SNAPSHOT_DIR}`
    * We reboot the machines as we do right now

The main difference is that an "innocent" reboot won't apply new snapshots randomly on the cluster, but we'll apply the new snapshots when the customer hits "Upgrade cluster" and they really want the cluster to be upgraded.

So, summing up: I see a combination of cluster-wide migrations and node-specific migrations. I think all of them should be idempotent just in case, as much as possible. We could do a "migrations-like" feature that only runs node migrations once per node. Also, we could be the ones applying the new btrfs snapshots instead of being `transactional-update` the one activating it automatically [making an "innocent" reboot issued by an operator load a new version without our control/knowledge].